File: DSA/PriyankaSingh_590016827/day38question3.c

#include <stdio.h>
#include <string.h>

#define MAX 100

int isRedundant(char *expr) {
    char stack[MAX];
    int top = -1;

    for (int i = 0; expr[i] != '\0'; i++) {
        if (expr[i] == ')') {
            int count = 0;
            while (top != -1 && stack[top] != '(') {
                char ch = stack[top--];
                if (ch == '+' || ch == '-' || ch == '*' || ch == '/')
                    count++;
            }
            top--; // pop '('
            if (count == 0)
                return 1;
        } else {
            stack[++top] = expr[i];
        }
    }
    return 0;
}

int main() {
    char expr[MAX];
    printf("Enter expression: ");
    scanf("%s", expr);

    if (isRedundant(expr))
        printf("Yes\n");
    else
        printf("No\n");

    return 0;
}

File: DSA/SakshamKashyap_590014183/day38/day38_Question_3.py
def find_redundant_brackets(expression):
    stack = []
    operators = {'+', '-', '*', '/'}

    for char in expression:
        if char == ')':
            is_redundant = True
            
            while stack and stack[-1] != '(':
                if stack.pop() in operators:
                    is_redundant = False
            
            stack.pop()
            if is_redundant:
                return "Yes"
        
        elif char == '(' or char in operators:
            stack.append(char)
            
    return "No"

# Test cases
print(find_redundant_brackets("((a+b))"))
print(find_redundant_brackets("(a+b)"))
